lib/Component.luau

```diff
@@ -43,28 +43,34 @@ local merge = require(script.Parent.immutable).merge
 -- This is a special value we set inside the component's metatable that will allow us to detect when
 -- a Component is accidentally inserted as a Component Instance.
 -- It should not be accessible through indexing into a component instance directly.
-local DIAGNOSTIC_COMPONENT_MARKER = {}
+local DIAGNOSTIC_COMPONENT_MARKER = table.freeze({})
+
+local PRIMITIVE_MARKER = table.freeze({})
 
 local lastId = 0
-local function newComponent(name, defaultData)
+local function new(name: string, defaultData)
 	name = name or debug.info(2, "s") .. "@" .. debug.info(2, "l")
 
-	assert(
-		defaultData == nil or type(defaultData) == "table",
-		"if component default data is specified, it must be a table"
-	)
+	local ComponentInstance = {}
+	ComponentInstance.__index = ComponentInstance
 
-	local component = {}
-	component.__index = component
+	function ComponentInstance.new(data)
+		local mt = getmetatable(ComponentInstance :: any)
+		if typeof(data) == "table" then
+			data = data or {}
 
-	function component.new(data)
-		data = data or {}
+			if defaultData then
+				data = merge(defaultData, data)
+			end
 
-		if defaultData then
-			data = merge(defaultData, data)
-		end
+			mt[PRIMITIVE_MARKER] = false
```

Comment:
This is redundant `mt[PRIMITIVE_MARKER] = false` since the absence of it will just be interpreted as false in the isPrimitive function already.